R/register_handlers.R

```diff
@@ -0,0 +1,98 @@
+#' Register handlers for logging messages, warnings and errors
+#'
+#' @param namespace (`character(1)`) the logger namespace
+#' @param package (`character(1)`) the package name
+#'
+#' @return `NULL` invisibly. Called for its side effects.
+#'
+#' @details This function registers global handlers for messages, warnings and errors.
+#' The handlers will investigate the call stack and if it contains a function
+#' from the package, the message, warning or error will be logged into the respective
+#' namespace.
+#' The handlers are registered only once per package and type. Consecutive calls will be ignored.
+#'
+#' @note Registering handlers is disallowed within `tryCatch` blocks which is used throughout
+#' base R and other packages. Because of this, the `register_handlers` function is wrapped in a
+#' `try` block to avoid errors.
+#'
+#' @seealso [globalCallingHandlers()]
+#'
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' register_logger(namespace = "teal.logger")
+#' register_handlers("teal.logger")
+#' # see the outcome
+#' globalCallingHandlers()
+#' }
+register_handlers <- function(namespace, package = namespace) {
+  try(
+    {
+      register_handler_type(namespace = namespace, package = package, type = "message")
+      register_handler_type(namespace = namespace, package = package, type = "warning")
+      register_handler_type(namespace = namespace, package = package, type = "error")
+    },
+    silent = TRUE
+  )
+
+  invisible(NULL)
+}
+
+#' @keywords internal
+register_handler_type <- function(
+    namespace,
+    package,
+    type = c("error", "warning", "message")) {
+  match.arg(type)
+  if (!(is.character(namespace) && length(namespace) == 1)) {
+    stop("namespace argument must be a scalar character.")
+  }
+  if (!(namespace %in% logger::log_namespaces())) {
+    stop("namespace argument must be a pre-registered logger namespace.")
+  }
+  if (!(is.character(package) && length(package) == 1)) {
+    stop("package argument must be a scalar character.")
+  }
+
+  logger_fun <- switch(type,
+    error = logger::log_error,
+    warning = logger::log_warn,
+    message = logger::log_info
+  )
+
+  # avoid re-registering the same handler
+  gch <- globalCallingHandlers()[names(globalCallingHandlers()) == type]
```

Comment:
any way you can use `match` to seek for the `type`?
```suggestion
  gch <- globalCallingHandlers()[match(type, names(globalCallingHandlers())]
```

Reply:
This is connected to one of your comment below. 
`match` returns `NA` on no-match, whereas `==` returns an empty vector instead. Note that we are using the outcome as a iterable in the next step. Iterating over NA might give ambiguous results. It is expected that we won't iterate when no-match.

Have a look at below example:
```
r$> for (i in letters[match("aa", letters)]) print("Hello!")
[1] "Hello!"
```
```
r$> for (i in letters[letters == "aa"]) print("Hello!")

```